packages/actor-query-operation-group/lib/ActorQueryOperationGroup.ts

```diff
@@ -0,0 +1,84 @@
+
+import { ArrayIterator } from 'asynciterator';
+import { termToString } from 'rdf-string';
+import { Algebra } from "sparqlalgebrajs";
+import { SimpleEvaluator } from 'sparqlee';
+
+import {
+  ActorQueryOperation,
+  ActorQueryOperationTypedMediated,
+  IActorQueryOperationOutputBindings,
+  IActorQueryOperationTypedMediatedArgs,
+} from "@comunica/bus-query-operation";
+import { ActionContext, IActorTest } from "@comunica/core";
+
+import { GroupsState } from './GroupsState';
+
+/**
+ * A comunica Group Query Operation Actor.
+ */
+export class ActorQueryOperationGroup extends ActorQueryOperationTypedMediated<Algebra.Group> {
+
+  constructor(args: IActorQueryOperationTypedMediatedArgs) {
+    super(args, 'group');
+  }
+
+  public async testOperation(pattern: Algebra.Group, context: ActionContext): Promise<IActorTest> {
+    for (const i in pattern.aggregates) {
+      // Will throw for unsupported expressions
+      const _ = new SimpleEvaluator(pattern.aggregates[i].expression);
+    }
+    return true;
+  }
+
+  public async runOperation(pattern: Algebra.Group, context: ActionContext)
+    : Promise<IActorQueryOperationOutputBindings> {
+
+    // Get result stream for the input query
+    const { input, aggregates } = pattern;
+    const outputRaw = await this.mediatorQueryOperation.mediate({ operation: input, context });
+    const output = ActorQueryOperation.getSafeBindings(outputRaw);
+
+    // The variables in scope are the variables on which we group, i.e. pattern.variables.
+    // For 'GROUP BY ?x, ?z', this is [?x, ?z], for 'GROUP by expr(?x) as ?e' this is [?e].
+    // But also in scope are the variables defined by the aggregations, since GROUP has to handle this.
+    const variables = pattern.variables
+      .map(termToString)
+      .concat(aggregates.map((agg) => termToString(agg.variable)));
+
+    // Return a new promise that completes when the stream has ended or when
+    // an error occurs
+    return new Promise((resolve, reject) => {
+      const groups = new GroupsState(pattern);
+
+      // Phase 2: Collect aggregator results
+      // We can only return when the binding stream ends, when that happens
+      // we return the identified groups. Which are nothing more than Bindings
+      // of the grouping variables merged with the aggregate variables
+      output.bindingsStream.on('end', () => {
```

Comment:
I would suggest moving phase 2 (the on-end listener) _after_ phase 1.
Unless this causes problems?

Reply:
This causes a race condition where the stream is completed and consumed before the the on-end listeners are bound.